lib/companies_web/repo_metrics_history.ex

```diff
@@ -0,0 +1,68 @@
+defmodule CompaniesWeb.RepoMetricsHistory do
+  @moduledoc """
+  GenServer to handle historical data for Ecto queries and rebroadcast to
+  LiveDashboard under seperate history-focused telemetry namespace.
+  """
+
+  use GenServer
+
+  @telemetry_event [:companies, :repo, :query]
+  @historic_metric [:ecto, :dashboard, :query]
+  @history_buffer_size 500
+
+  def signatures do
+    %{
+      @historic_metric => {__MODULE__, :data, []}
+    }
+  end
+
+  def telemetry_event, do: @telemetry_event
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
+  end
+
+  def init(_state) do
+    {:ok, %{history: CircularBuffer.new(@history_buffer_size)}}
+  end
+
+  def data(%{event_name: event_name} = metric) do
+    if List.starts_with?(event_name, @historic_metric) do
+      GenServer.call(__MODULE__, {:data, metric})
+    else
+      []
+    end
+  end
+
+  def setup_handlers do
+    :telemetry.attach(
+      "aggregation-handler-#{__MODULE__}",
+      @telemetry_event,
+      &__MODULE__.handle_event/4,
+      nil
+    )
+  end
+
+  def handle_event(@telemetry_event, metric_map, metadata, config) do
+    GenServer.cast(__MODULE__, {:telemetry_metric, metric_map, metadata, config})
+  end
+
+  def handle_call({:data, _metric}, _from, %{history: history}) do
+    {:reply, CircularBuffer.to_list(history), %{history: history}}
+  end
+
+  def handle_cast({:telemetry_metric, metric_map, metadata, _config}, %{history: history}) do
+    time = System.system_time(:second)
+    :telemetry.execute(@historic_metric, metric_map, metadata)
+
+    new_history = CircularBuffer.insert(history, %{data: metric_map, time: time, metadata: pruned_metadata(metadata)})
+
+    {:noreply, %{history: new_history}}
+  end
+
+  defp pruned_metadata(metadata) do
+    # for now keep it all, reminder to either keep or drop selected fields based on dashboard usage to conserve memory,
+    # ideally via some published source of truth hook from dahsboard module to ensure correctness
+    metadata
```

Comment:
José expressed concern about memory, and I think this will be the main way we address that concern in practice, though this method will likely need to consider the metric as well as the metadata to decide what to save vs discard... but I think for now we can either leave this as is, or kill, and look at memory usage without discarding anything before deciding if it's a practical issue for us, though we may want to tackle even if it's not a big issue here so that visibility is in place in LiveDashboard to support hooks doing this pruning intelligently.

Reply:
On newer versions of my fork, I'm actually just leaving out the metadata entirely and most of the metrics seem to work exactly the same, so possibly we can do the same here, though I'm curious to see the difference in memory impact both ways/it's entirely possible updates to LiveDashboard may start making use of this metadata in new releases that it's not using now.